src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -449,13 +456,22 @@ async def fetch_with_coursier(request: CoursierFetchRequest) -> FallibleClasspat
     # TODO: Loading this per JvmArtifact.
     lockfile = await Get(CoursierResolvedLockfile, CoursierResolveKey, request.resolve)
 
+    requirement = ArtifactRequirement.from_jvm_artifact_target(request.component.representative)
+
+    if lockfile.metadata and not lockfile.metadata.is_valid_for([requirement]):
+        raise ValueError(
```

Comment:
`ValueError` seems to be the exception we've been using in this part of the codebase. Let me know if there's anything more appropriate here.

Reply:
We could probably define a new error type, but NBD.